docs/design/kata-nydus-design.md

```diff
@@ -0,0 +1,89 @@
+
+The following benchmarking result shows the performance improvement compared with the OCI image for the container cold startup elapsed time on containerd. As the OCI image size increases, the container startup time of using Nydus image remains very short.
+
+![perf](arch-images/perf.jpeg)
+
+## Proposal - Bring lazyload ability to Kata Containers
+
+Nydusd is a fuse/virtiofs daemon which is provided by nydus project and it supports Passthroughfs and [RAFS](https://github.com/dragonflyoss/image-service/blob/master/docs/nydus-design.md) (Registry Acceleration File System) natively, so in Kata Containers, we can use nydusd in place of virtiofsd and mount nydus image to guest in the meanwhile.
+
+The process of create/start Kata Containers with virtiofsd,
+
+1. when create sandbox, the Kata Containers shim will launch virtiofsd before VM start, so VM can share dir with host.
+2. when create container, the Kata Containers shim will mount rootfs to /path/to/bundle/rootfs and bind mount rootfs to `sharedir`, so it can be seen in the guest.
+
+The process of create/start Kata Containers with nydusd,
+
+![kata-nydus](arch-images/kata-nydus.png)
+
+1. when create sandbox, the Kata Containers Containerd v2 shim start `nydusd` daemon and mount `passthroughfs` filesystem before the VM starts.
+
+```bash
+# start nydusd
+$ sudo /usr/local/bin/nydusd --log-level info  --sock /run/vc/vm/5502d6f9a404ccca94320d7d47f9225e25e05d193d521d713565c01811e480d5/vhost-user-fs.sock --apisock  /run/vc/vm/5502d6f9a404ccca94320d7d47f9225e25e05d193d521d713565c01811e480d5/api.sock
+```
+
+```bash
+# mount passthroughfs
+$ curl -v --unix-socket /run/vc/vm/5502d6f9a404ccca94320d7d47f9225e25e05d193d521d713565c01811e480d5/api.sock \
+    -X POST "http://localhost/api/v1/mount?mountpoint=/foo1" -H 'accept: */*' \
+    -H "Content-Type: application/json" \
+    -d '{
+            "source":"/path/to/sharedir",
+            "fs_type":"passthrough_fs",
+            "config":""
+    }'
+```
+
+2. when create normal container, the Kata Containers Containerd v2 shim send request to nydusd to mount `rafs` to guest. The Kata Containers Containerd v2 shim will also bind mount `snapshotdir` which `nydus-snapshotter` assigns to `sharedir`, so in guest, container rootfs=overlay(lowerdir=`rafs`, upperdir=`snapshotdir/fs`, workdir=`snapshotdir/work`)
+
+```bash
+# mount rafs
+# source: the metafile of nydus image
+# config: the config of this image
+$ curl --unix-socket /run/vc/vm/5502d6f9a404ccca94320d7d47f9225e25e05d193d521d713565c01811e480d5/api.sock \
+    -X POST "http://localhost/api/v1/mount?mountpoint=/bar1" -H 'accept: */*' \
+    -H "Content-Type: application/json" \
+    -d '{
+            "source":"/path/to/bootstrap",
+            "fs_type":"rafs",
+            "config":"{}",
+    }'
+```
+
+> So the question is `how to transfer the rafs info from nydus-snapshotter to the Kata Containers Containerd v2 shim?`
```

Comment:
This line is a TODO?

Reply:
No, it is not a TODO, it has been implemented in this pr with nydus

Reply:
So the `So the question is ...` should be deleted?

Reply:
yeah, done